### PR TITLE
Add a foreign key relation from process instance to process definition

### LINKED ISSUE #13495: Add a foreign key relation from process instance to process definition
**Description**

Recently, we encountered [a critical bug](https://github.com/camunda/zeebe/issues/13254) where process instances were created for a process definition that was not deployed successfully.

While that bug came to be due to problems with cache invalidation, it struck me as odd that we don't have a relational data constraint between process instances and the process definition they belong to.

With #13475 merged, we already have a place where this relation is stored. But, the foreign key relation is not explicitly defined and thus not asserted.

We should consider the performance/safety trade-off to defining this as a foreign key relationship that can be asserted as part of the data consistency checks feature flag. AFAIK, the performance cost is one additional lookup per insertion at one insertion per process instance.

See: 
- https://github.com/camunda/zeebe/issues/13254#issuecomment-1617550522
- https://github.com/camunda/zeebe/pull/13373#discussion_r1261633196

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

This PR add a foreign key between process instances and the process definition they belong to

## Related issues

closes #13495 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by remcowesterhoud — COMMENTED** [review]
Thanks Nic. Personally I'd prefer to not merge this yet as this will conflict with the changes for multi-tenancy.

I'm also missing a testcase which verifies that an exception is thrown when the foreign key constraint is violated.

**Comment by nicpuppa** [comment]
> Thanks Nic. Personally I'd prefer to not merge this yet as this will conflict with the changes for multi-tenancy.

I'm agree with you, if you see conflict with multitenancy we can wait.
> 
> I'm also missing a testcase which verifies that an exception is thrown when the foreign key constraint is violated.

I can add a test case for this

Thank you @remcowesterhoud 💪

**Comment by nicpuppa** [comment]
Hey @remcowesterhoud this PR is still valid for you ? Wdyt ?

**Comment by npepinpe** [comment]
Considering Remco left, what's the state here?

**Comment by korthout** [comment]
@nicpuppa Considering [the benchmark results](https://github.com/camunda/zeebe/pull/14105#issuecomment-1702566068), I wonder whether we should proceed with this. What do you think? Was this measurement reliable?

**Comment by Zelldon** [comment]
@megglos @nicpuppa what is the state here?

**Comment by megglos** [comment]
@nicpuppa will try to fix the performance impact or actually close the PR

**Comment by oleschoenburg** [comment]
The performance impact is just when consistency checks are enabled, right? The linked benchmark results only shows that the network latency impact is flaky, unfortunately :-/

**Comment by nicpuppa** [comment]
I rebased the branch to the current main, and added multi tenancy and PIT support. I still need to run a benchmark to understand the impact of the fk addition

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/ElementInstanceStateTest.java
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/ElementInstanceStateTest.java]: 🔧 This method adds a process definition, not an instance. Let's adjust the name

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: ❓ Why does this now have to be tenant-specific? The column family currently also exists as non-tenant specific. Note that the process instance and process definition keys are already fully unique, and their referenced values are tenant-specific.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @nicpuppa

❌ I have some questions regarding the tenant-specific code. I wonder whether it's necessary. Please have a look.

💭 Curious to the benchmark results.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java
Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Because with this changes I added a fk to the process definition. The key of `processInstanceKeyByProcessDefinitionKeyColumnFamily` now contain also the tenant id. Is this make sense to you ?

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: 💭 Ah I see now. The `PROCESS_CACHE` column family contains the process definitions as a tenant aware key (tenant id is prefixed). So, in order for the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family to reference the `PROCESS_CACHE` we must also prefix the `processDefinitionKey` in the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family.

❌ This means we need to migrate the existing data in the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family and prefix it in a similar way as we did with the multi-tenancy migrations.

Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: I'd strongly suggest not migrating data this way, especially if it's not strictly required. We already had issues with the multi-tenancy migrations and adding another one is only going to make this worse.

Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Yes but there should almost always be ways to avoid data migrations, for example by being backwards compatible with old schemas, by migrating data incrementally as needed or other strategies.
While we did make improvements to he way we run migrations to unblock us, these migrations inevitably require a lot of resources for large runtime use cases because our migrations are (at least) linear in the amount of data, here it would be the number of process instances.

> If problems are left, what can we do to overcome those problems in the future?
I'd suggest using a different strategy to implement this feature. For example (and if I understand this PR correctly), we could implement it such that new entries _are_ written with the tenant reference while old entries either do not benefit from the consistency checks or have to check all tenants for a matching process instance.

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: >new entries are written with the tenant reference while old entries either do not benefit from the consistency checks or have to check all tenants for a matching process instance.

That sounds hard to maintain, and we explicitly chose not to take this route with multi-tenancy for that reason. The benefit of migrating the data is to avoid having to maintain two different ways of storing the data.

I guess the trade-off here is (pick 2): 
1. update performance, i.e. shorter or longer downtime of a partition during update
2. data consistency safety for process instance -> process definition
3. state code maintainability

Not sure what's best though. My gut says 2 and 3, as 1 is a one-time downside/upside, but 2 and 3 are both long-term.

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Since with 8.5 we run migration only on version upgrade, I think adding a new one has not a big impact for customer on daily basis. As already mentioned, the impact is on version upgrading. Might be worth to invest more time in making migration even more "smart". For this case, a migration should run only if updating from 8.5.x to 8.6.x

In general, I'm in favour of a longer downtime if we improve data consistency

Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: > update performance, i.e. shorter or longer downtime of a partition during update

Past experience was not that the update was "slow" but that it crashed brokers. At the very least I'd suggest you test new migrations on some large runtime state setups to ensure that they can finish within some memory bound.

Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Am I right that we want to run another round of migrations for all existing PIs for introducing this FK which by default has no effect as consistency checks are disabled by default? These migrations will affect all users and not just those actually using consistency checks, right?

If so I would conclude the cost-benefit of this change is actually not worth doing the change at all and we rather not introduce the FK. The only compromise I see is implementing this check without a migration as consistency checks have a significant performance check anyway.

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: >has no effect as consistency checks are disabled by default

Might make sense to reverse that at some point, making data safety the default, and allowing users to opt-out for performance

Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: > In general, I'm in favour of a longer downtime if we improve data consistency

By default users will get the downtime but not the latter, right? Also we embrace here a potential large migration transaction state risking to hit https://github.com/camunda/zeebe/issues/14975 for all users of which only a fraction has consistency checks and MT enabled.

I get we at some point have no other option but to migrate but in the context of this change I see more risks and downsides for the majority of users with a migration compared to implementing the check with a higher performance impact for a smaller subset of users that don't seem to have high performance demands anyway if checks are enabled.

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Ok, so as we don't yet have consistency checks enabled by default, I agree that we shouldn't add this foreign key because the downside (downtime during upgrade) doesn't outweigh the upside (data safety only to those that have consistency checks enabled, i.e. few).

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Should I close this PR then ?

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java:48]: Ok with me. What about you, @megglos?

**Pull request closed without merging**
